File: backend-rust/src/core/parser.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use regex::Regex;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AggregateSpec {
    pub func: String,
    pub expression: String,
    pub alias: String,
}

impl AggregateSpec {
    pub fn is_count_star(&self) -> bool {
        self.func.eq_ignore_ascii_case("count") && self.expression == "*"
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderBySpec {
    pub key: String,
    pub descending: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParsedQuery {
    pub raw_sql: String,
    pub table: String,
    pub select_items: Vec<String>,
    pub aggregates: Vec<AggregateSpec>,
    pub group_by: Vec<String>,
    pub where_clause: Option<String>,
    pub order_by: Vec<OrderBySpec>,
    pub limit: Option<usize>,
}

impl ParsedQuery {
    pub fn projection_columns(&self) -> Vec<String> {
        let mut columns = self.group_by.clone();

        let mut expression_sources = self.group_by.clone();

        if let Some(where_clause) = &self.where_clause {
            expression_sources.push(where_clause.clone());
        }

        for aggregate in &self.aggregates {
            if aggregate.expression != "*" {
                expression_sources.push(aggregate.expression.clone());
            }
        }

        for expression in expression_sources {
            for identifier in extract_identifiers(&expression) {
                if !columns.contains(&identifier) {
                    columns.push(identifier);
                }
            }
        }

        columns
    }
}

pub const SUPPORTED_AGGREGATES: &[&str] = &[
    "count",
    "sum",
    "avg",
];

pub const SQL_KEYWORDS: &[&str] = &[
    "and",
    "as",
    "asc",
    "avg",
    "between",
    "by",
    "coalesce",
    "count",
    "date",
    "desc",
    "from",
    "group",
    "in",
    "is",
    "limit",
    "like",
    "not",
    "null",
    "or",
    "order",
    "select",
    "sum",
    "where",
];

pub fn split_top_level_csv(text: &str) -> Vec<String> {
    let mut items = Vec::new();
    let mut depth = 0;
    let mut current = String::new();

    for ch in text.chars() {
        match ch {
            '(' => {
                depth += 1;
                current.push(ch);
            }
            ')' => {
                depth -= 1;
                current.push(ch);
            }
            ',' if depth == 0 => {
                items.push(current.trim().to_string());
                current.clear();
            }
            _ => current.push(ch),
        }
    }

    let tail = current.trim();

    if !tail.is_empty() {
        items.push(tail.to_string());
    }

    items
}

pub fn normalize_alias(func: &str, expression: &str) -> String {
    let target = if expression == "*" {
        "all".to_string()
    } else {
        expression
            .chars()
            .map(|c| {
                if c.is_ascii_alphanumeric() || c == '_' {
                    c
                } else {
                    '_'
                }
            })
            .collect::<String>()
            .trim_matches('_')
            .to_lowercase()
    };

    format!("{}_{}", func.to_lowercase(), target)
}

pub fn extract_identifiers(expression: &str) -> Vec<String> {
    let string_re =
        Regex::new(r"'(?:''|[^'])*'").unwrap();

    let ident_re =
        Regex::new(r"[a-zA-Z_][a-zA-Z0-9_]*").unwrap();

    let cleaned =
        string_re.replace_all(expression, " ");

    let mut seen = HashSet::new();
    let mut identifiers = Vec::new();

    for capture in ident_re.find_iter(&cleaned) {
        let token = capture.as_str();

        if SQL_KEYWORDS
            .iter()
            .any(|kw| kw.eq_ignore_ascii_case(token))
        {
            continue;
        }

        if seen.insert(token.to_string()) {
            identifiers.push(token.to_string());
        }
    }

    identifiers
}

pub fn parse_aggregate(item: &str) -> Result<AggregateSpec, String> {
    let aggregate_start =
        Regex::new(r"(?i)^(count|sum|avg)\s*\(")
            .unwrap();

    let captures = aggregate_start
        .captures(item.trim())
        .ok_or_else(|| {
            format!(
                "Unsupported select expression: {}",
                item
            )
        })?;

    let func = captures
        .get(1)
        .unwrap()
        .as_str()
        .to_lowercase();

    if !SUPPORTED_AGGREGATES.contains(&func.as_str()) {
        return Err(format!(
            "Unsupported aggregate function: {}",
            func
        ));
    }

    let whole_match = captures.get(0).unwrap();

    let remainder =
        &item.trim()[whole_match.end()..];

    let mut depth = 1;
    let mut closing_index = None;

    for (idx, ch) in remainder.char_indices() {
        match ch {
            '(' => depth += 1,
            ')' => {
                depth -= 1;

                if depth == 0 {
                    closing_index = Some(idx);
                    break;
                }
            }
            _ => {}
        }
    }

    let closing_index =
        closing_index.ok_or_else(|| {
            format!(
                "Malformed aggregate expression: {}",
                item
            )
        })?;

    let expression =
        remainder[..closing_index].trim();

    let tail =
        remainder[closing_index + 1..].trim();

    let alias_re = Regex::new(
        r"(?i)^(?:as\s+)?([a-zA-Z_][a-zA-Z0-9_]*)?$",
    )
    .unwrap();

    let alias_capture = alias_re
        .captures(tail)
        .ok_or_else(|| {
            format!(
                "Unsupported select expression: {}",
                item
            )
        })?;

    let alias = alias_capture
        .get(1)
        .map(|m| m.as_str().to_string())
        .unwrap_or_else(|| {
            normalize_alias(
                &func,
                expression,
            )
        });

    Ok(AggregateSpec {
        func,
        expression: expression.to_string(),
        alias,
    })
}


pub fn parse_order_by(
    order_by_clause: Option<&str>,
    parsed: Option<&ParsedQuery>,
) -> Result<Vec<OrderBySpec>, String> {
    let Some(clause) = order_by_clause else {
        return Ok(vec![]);
    };

    let mut specs = Vec::new();

    let re = Regex::new(
        r"(?i)^(.+?)(?:\s+(asc|desc))?$"
    )
    .unwrap();

    for item in split_top_level_csv(clause) {
        let captures = re
            .captures(item.trim())
            .ok_or_else(|| {
                format!(
                    "Unsupported ORDER BY expression: {}",
                    item
                )
            })?;

        let mut key = captures
            .get(1)
            .unwrap()
            .as_str()
            .trim()
            .to_string();

        if let Some(parsed) = parsed {
            for aggregate in &parsed.aggregates {
                let signature = format!(
                    "{}({})",
                    aggregate.func,
                    aggregate.expression
                );

                if key
                    .replace(" ", "")
                    .eq_ignore_ascii_case(
                        &signature.replace(" ", "")
                    )
                {
                    key = aggregate.alias.clone();
                    break;
                }
            }
        }

        let descending = captures
            .get(2)
            .map(|m| {
                m.as_str()
                    .eq_ignore_ascii_case("desc")
            })
            .unwrap_or(false);

        specs.push(OrderBySpec {
            key,
            descending,
        });
    }

    Ok(specs)
}

pub fn parse_analytical_query(
    query: &str,
) -> Result<ParsedQuery, String> {
    let normalized = query
        .trim()
        .trim_end_matches(';')
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");

    let query_re = Regex::new(
        r"(?i)^select\s+(?P<select>.+)\s+from\s+(?P<table>[a-zA-Z_][a-zA-Z0-9_]*)(?:\s+where\s+(?P<where>.*?))?(?:\s+group\s+by\s+(?P<group_by>.*?))?(?:\s+order\s+by\s+(?P<order_by>.*?))?(?:\s+limit\s+(?P<limit>\d+))?$"
    ).unwrap();

    let captures = query_re
        .captures(&normalized)
        .ok_or_else(|| {
            "Approx mode supports SELECT aggregate queries on one table with optional WHERE/GROUP BY/ORDER BY/LIMIT"
                .to_string()
        })?;

    let select_items = split_top_level_csv(
        captures.name("select").unwrap().as_str()
    );

    let mut aggregates = Vec::new();
    let mut plain_columns = Vec::new();

    let aggregate_start = Regex::new(r"(?i)^(count|sum|avg)\s*\(").unwrap();

    for item in &select_items {
        if aggregate_start.is_match(item.trim()) {
            aggregates.push(parse_aggregate(item)?);
        } else {
            plain_columns.push(item.trim().to_string());
        }
    }

    if aggregates.is_empty() {
        return Err(
            "Approx mode requires at least one aggregate expression".to_string(),
        );
    }

    let group_by = captures
        .name("group_by")
        .map(|m| split_top_level_csv(m.as_str()))
        .unwrap_or_default();

    let normalized_plain: Vec<String> = plain_columns
        .iter()
        .map(|s| s.to_lowercase())
        .collect();

    let normalized_group: Vec<String> = group_by
        .iter()
        .map(|s| s.to_lowercase())
        .collect();

    if normalized_plain != normalized_group {
        return Err(
            "Non-aggregate SELECT columns must match GROUP BY columns in order".to_string(),
        );
    }

    let mut parsed = ParsedQuery {
        raw_sql: normalized.clone(),
        table: captures
            .name("table")
            .unwrap()
            .as_str()
            .to_string(),
        select_items,
        aggregates,
        group_by,
        where_clause: captures
            .name("where")
            .map(|m| m.as_str().trim().to_string()),
        order_by: vec![],
        limit: captures
            .name("limit")
            .map(|m| m.as_str().parse::<usize>().unwrap()),
    };

    parsed.order_by = parse_order_by(
        captures.name("order_by").map(|m| m.as_str()),
        Some(&parsed),
    )?;

    Ok(parsed)
}

File: backend/core/parser.py
from __future__ import annotations

import re
from dataclasses import dataclass


SUPPORTED_AGGREGATES = {"count", "sum", "avg"}
SQL_KEYWORDS = {
    "and",
    "as",
    "asc",
    "avg",
    "between",
    "by",
    "coalesce",
    "count",
    "date",
    "desc",
    "from",
    "group",
    "in",
    "is",
    "limit",
    "like",
    "not",
    "null",
    "or",
    "order",
    "select",
    "sum",
    "where",
}


@dataclass(frozen=True)
class AggregateSpec:
    func: str
    expression: str
    alias: str

    @property
    def is_count_star(self) -> bool:
        return self.func == "count" and self.expression == "*"


@dataclass(frozen=True)
class OrderBySpec:
    key: str
    descending: bool = False


@dataclass(frozen=True)
class JoinSpec:
    join_type: str  # INNER, LEFT, RIGHT, FULL
    right_table: str
    on_condition: str


@dataclass(frozen=True)
class ParsedQuery:
    raw_sql: str
    table: str
    select_items: list[str]
    aggregates: list[AggregateSpec]
    group_by: list[str]
    where_clause: str | None = None
    order_by: list[OrderBySpec] | None = None
    limit: int | None = None
    joins: list[JoinSpec] | None = None

    @property
    def original_query(self) -> str:
        return self.raw_sql

    @property
    def has_joins(self) -> bool:
        return self.joins is not None and len(self.joins) > 0

    @property
    def all_tables(self) -> list[str]:
        """Return all tables involved in the query (base + joined tables)"""
        tables = [self.table]
        if self.joins:
            tables.extend([join.right_table for join in self.joins])
        return tables

    @property
    def projection_columns(self) -> list[str]:
        columns = list(dict.fromkeys(self.group_by))
        expression_sources = [*self.group_by]
        if self.where_clause:
            expression_sources.append(self.where_clause)
        for aggregate in self.aggregates:
            if aggregate.expression != "*":
                expression_sources.append(aggregate.expression)
        for expression in expression_sources:
            for identifier in _extract_identifiers(expression):
                if identifier not in columns:
                    columns.append(identifier)
        return columns


def _split_top_level_csv(text: str) -> list[str]:
    items: list[str] = []
    depth = 0
    current: list[str] = []

    for char in text:
        if char == "(":
            depth += 1
        elif char == ")":
            depth -= 1
        elif char == "," and depth == 0:
            items.append("".join(current).strip())
            current = []
            continue
        current.append(char)

    tail = "".join(current).strip()
    if tail:
        items.append(tail)
    return items


def _normalize_alias(func: str, expression: str) -> str:
    if expression == "*":
        target = "all"
    else:
        target = re.sub(r"[^a-zA-Z0-9_]", "_", expression).strip("_").lower()
    return f"{func.lower()}_{target}"


def _parse_aggregate(item: str) -> AggregateSpec:
    match = re.match(r"(?is)^(count|sum|avg)\s*\(", item.strip())
    if not match:
        raise ValueError(f"Unsupported select expression: {item}")

    func = match.group(1).lower()
    remainder = item.strip()[match.end() :]
    depth = 1
    closing_index = None
    for index, char in enumerate(remainder):
        if char == "(":
            depth += 1
        elif char == ")":
            depth -= 1
            if depth == 0:
                closing_index = index
                break
    if closing_index is None:
        raise ValueError(f"Malformed aggregate expression: {item}")

    expression = remainder[:closing_index].strip()
    tail = remainder[closing_index + 1 :].strip()
    alias_match = re.match(r"(?is)^(?:as\s+)?([a-zA-Z_][a-zA-Z0-9_]*)?$", tail)
    if not alias_match:
        raise ValueError(f"Unsupported select expression: {item}")
    alias = alias_match.group(1) or _normalize_alias(func, expression)
    if func not in SUPPORTED_AGGREGATES:
        raise ValueError(f"Unsupported aggregate function: {func}")

    return AggregateSpec(func=func, expression=expression, alias=alias)


def _extract_identifiers(expression: str) -> list[str]:
    cleaned = re.sub(r"'(?:''|[^'])*'", " ", expression)
    candidates = re.findall(r"[a-zA-Z_][a-zA-Z0-9_]*", cleaned)
    identifiers: list[str] = []
    for token in candidates:
        if token.lower() in SQL_KEYWORDS:
            continue
        if token not in identifiers:
            identifiers.append(token)
    return identifiers


def _parse_order_by(order_by_clause: str | None, parsed: ParsedQuery | None = None) -> list[OrderBySpec]:
    if not order_by_clause:
        return []

    order_specs: list[OrderBySpec] = []
    for item in _split_top_level_csv(order_by_clause):
        match = re.match(r"(?is)^(.+?)(?:\s+(asc|desc))?$", item.strip())
        if not match:
            raise ValueError(f"Unsupported ORDER BY expression: {item}")

        key = match.group(1).strip()
        if parsed is not None:
            for aggregate in parsed.aggregates:
                agg_signature = f"{aggregate.func}({aggregate.expression})"
                if key.lower().replace(" ", "") == agg_signature.replace(" ", ""):
                    key = aggregate.alias
                    break

        order_specs.append(OrderBySpec(key=key, descending=(match.group(2) or "").lower() == "desc"))

    return order_specs


def _parse_joins(query_from_onwards: str, base_table: str) -> tuple[list[JoinSpec], str]:
    """
    Extract JOIN clauses from query string.
    Returns (list of JoinSpecs, remaining query after joins)
    """
    joins: list[JoinSpec] = []
    remaining = query_from_onwards

    # Pattern to match JOIN clauses: [INNER|LEFT|RIGHT|FULL] JOIN table ON condition
    join_pattern = re.compile(
        r"(?is)\s+((?:inner|left|right|full(?:\s+outer)?)\s+)?join\s+"
        r"([a-zA-Z_][a-zA-Z0-9_]*)\s+on\s+(.+?)(?=\s+(?:inner|left|right|full|join|where|group|order|limit|$))",
    )

    for match in join_pattern.finditer(remaining):
        join_type = (match.group(1) or "inner").strip().upper()
        if "OUTER" in join_type:
            join_type = join_type.replace("OUTER", "").strip()
        right_table = match.group(2).strip()
        on_condition = match.group(3).strip()

        joins.append(JoinSpec(
            join_type=join_type,
            right_table=right_table,
            on_condition=on_condition,
        ))

    # Remove all JOIN clauses from remaining query
    remaining = join_pattern.sub("", remaining)

    return joins, remaining


def parse_analytical_query(query: str) -> ParsedQuery:
    normalized = query.strip().rstrip(";")

    # Enhanced pattern to capture optional JOIN clauses
    match = re.match(
        r"(?is)^select\s+(?P<select>.+?)\s+from\s+(?P<table>[a-zA-Z_][a-zA-Z0-9_]*)"
        r"(?P<remainder>.*?)$",
        normalized,
    )
    if not match:
        raise ValueError(
            "Approx mode supports SELECT aggregate queries with optional JOINs/WHERE/GROUP BY/ORDER BY/LIMIT"
        )

    base_table = match.group("table")
    remainder = match.group("remainder")

    # Parse JOINs first
    joins, remainder = _parse_joins(remainder, base_table)

    # Now parse WHERE/GROUP BY/ORDER BY/LIMIT from the remainder
    clause_match = re.match(
        r"(?is)^\s*"
        r"(?:where\s+(?P<where>.+?))?"
        r"(?:\s+group\s+by\s+(?P<group_by>.+?))?"
        r"(?:\s+order\s+by\s+(?P<order_by>.+?))?"
        r"(?:\s+limit\s+(?P<limit>\d+))?\s*$",
        remainder,
    )

    if not clause_match:
        raise ValueError(
            "Invalid SQL syntax in WHERE/GROUP BY/ORDER BY/LIMIT clauses"
        )

    select_items = _split_top_level_csv(match.group("select"))
    aggregates: list[AggregateSpec] = []
    plain_columns: list[str] = []
    for item in select_items:
        if re.match(r"(?is)^(count|sum|avg)\s*\(", item.strip()):
            aggregates.append(_parse_aggregate(item))
        else:
            plain_columns.append(item.strip())

    if not aggregates:
        raise ValueError("Approx mode requires at least one aggregate expression")

    group_by = _split_top_level_csv(clause_match.group("group_by")) if clause_match.group("group_by") else []
    normalized_plain = [column.lower() for column in plain_columns]
    normalized_group = [column.lower() for column in group_by]
    if normalized_plain != normalized_group:
        raise ValueError("Non-aggregate SELECT columns must match GROUP BY columns in order")

    parsed = ParsedQuery(
        raw_sql=normalized,
        table=base_table,
        select_items=select_items,
        aggregates=aggregates,
        group_by=group_by,
        where_clause=clause_match.group("where").strip() if clause_match.group("where") else None,
        limit=int(clause_match.group("limit")) if clause_match.group("limit") else None,
        joins=joins if joins else None,
    )
    order_by = _parse_order_by(clause_match.group("order_by"), parsed=parsed)
    return ParsedQuery(
        raw_sql=parsed.raw_sql,
        table=parsed.table,
        select_items=parsed.select_items,
        aggregates=parsed.aggregates,
        group_by=parsed.group_by,
        where_clause=parsed.where_clause,
        order_by=order_by,
        limit=parsed.limit,
        joins=parsed.joins,
    )
